src/hypofuzz/provider.py

```diff
@@ -26,29 +80,409 @@ def fresh_choice(
     return cast(ChoiceT, choice)
 
 
+# like choices_size, but handles ChoiceTemplate
+def _choices_size(choices: tuple[Union[ChoiceT, ChoiceTemplate], ...]) -> int:
+    return sum(
+        1 if isinstance(choice, ChoiceTemplate) else choices_size([choice])
+        for choice in choices
+    )
+
+
+class ReplayPriority(IntEnum):
+    FAILURE_SHRUNK = 0
+    FAILURE_UNSHRUNK = 1
+    COVERING = 2
+
+
+@dataclass
+class State:
+    choices: ChoicesT
+    replay_priority: Optional[ReplayPriority]
+    start_time: float
+    branches: Optional[frozenset[Branch]] = None
+    observation: Optional[Observation] = None
+    extra_queue_data: Optional[Any] = None
+
+
+# (priority, choice sequence, extra data)
+ReplayQueueElement = tuple[ReplayPriority, ChoicesT, Optional[Any]]
+
+
 class HypofuzzProvider(PrimitiveProvider):
+    add_observability_callback: ClassVar[bool] = True
+
     def __init__(
-        self, conjecturedata: Optional[ConjectureData], /, *, choices: ChoicesT
+        self,
+        conjecturedata: Optional[ConjectureData],
+        /,
+        # allow test-time override of the coverage collector.
+        collector: Optional[CoverageCollector] = None,
     ) -> None:
         super().__init__(conjecturedata)
-        self.choices = choices
-        self.index = 0
+        self.collector = collector
+        self._index = 0
+        self._replay_queue: SortedList[ReplayQueueElement] = SortedList(
+            key=lambda x: (x[0], _choices_size(x[1]))
+        )
+
+        # TODO: thread through a random instance from FuzzProcess here, or access
+        # the current BuildContext random?
+        self.random = Random()
+        self.ninputs = 0
+        self.elapsed_time = 0.0
+        self.since_new_branch = 0
+        self.phase: Optional[Phase] = None
+        self.status_counts = StatusCounts()
+
+        self._last_report: Report | None = None
+        self._last_saved_report_at = -math.inf
+        self._last_observed = -math.inf
+        self._started = False
+
+        self.worker_identity: Optional[WorkerIdentity] = None
+        self.database_key: Optional[bytes] = None
+        self.corpus: Optional[Corpus] = None
+        self.db: Optional[HypofuzzDatabase] = None
+        # per-test-case state, reset at the beginning of each test case.
+        self._state: Optional[State] = None
+
+    def _startup(self) -> None:
+        assert (
+            settings().database is not None
+        ), "pointless to use HypofuzzProvider without a database"
```

Comment:
probably want an if/raise here, just in case someone runs their fuzzer with `python -O ...`